src/ansys/fluent/core/session.py

```diff
@@ -343,8 +346,82 @@ def download(self, file_name: str, local_file_name: Optional[str] = None):
         FileNotFoundError
             If the remote file does not exist.
         """
-        if self.file_service:
-            if self.file_service.file_exist(file_name):
-                self.file_service.download_file(file_name, local_file_name)
+        self._upload_download_helper(
+            is_upload=False, file_name=file_name, local_file_name=local_file_name
+        )
+
+    def _wait_for_file(self, file_name: str):
+        """Wait for file to get ready for upload or download.
+
+        Parameters
+        ----------
+        file_name : str
+            File name
+        Raises
+        ------
+        FileNotFoundError
+            If a file does not exist.
+        """
+        start_time = time.time()
+        max_wait_time = sys.maxsize
+        while (time.time() - start_time) < max_wait_time:
+            if self.file_service.file_exist(os.path.basename(file_name)):
+                break
+            max_wait_time -= 1
+            time.sleep(3)
+        else:
+            raise FileNotFoundError(f"{file_name} does not exist.")
+
+    def _pypim_upload_download_helper(
+        self,
+        is_upload: bool,
+        file_name: str,
+        api: Optional[Any] = None,
```

Comment:
👍 for making this optional. Since it can be None, we should check it before calling it.

Can we separate out the upload and download helpers:
```
def _pypim_upload(
        self,
        file_name: str,
        on_uploaded: Optional[Any] = None):
    ...

Reply:
* Added check for `on_uploaded`.
* Separated pypim upload and download.

Thank you.

Reply:
* Called callback before the download.